vyper/semantics/analysis/imports.py

```diff
@@ -0,0 +1,328 @@
+        if module_ast in self._path:
+            cycle = self._path + [module_ast]
+            raise ImportCycle(" imports ".join(f'"{t.path}"' for t in cycle))
+
+        self._path.append(module_ast)
+        self._imports.append({})
+
+    def pop_path(self, expected: vy_ast.Module) -> None:
+        popped = self._path.pop()
+        if expected != popped:
+            raise CompilerPanic("unreachable")
+        self._imports.pop()
+
+    @contextlib.contextmanager
+    def enter_path(self, module_ast: vy_ast.Module) -> Iterator[None]:
+        self.push_path(module_ast)
+        try:
+            yield
+        finally:
+            self.pop_path(module_ast)
+
+
+class ImportAnalyzer:
+    def __init__(self, input_bundle: InputBundle, graph: _ImportGraph):
+        self.input_bundle = input_bundle
+        self.graph = graph
+        self._ast_of: dict[int, vy_ast.Module] = {}
+
+        self.seen: set[int] = set()
+
+        self.integrity_sum = None
+
+    def resolve_imports(self, module_ast: vy_ast.Module):
+        self._resolve_imports_r(module_ast)
+        self.integrity_sum = self._calculate_integrity_sum_r(module_ast)
+
+    def _calculate_integrity_sum_r(self, module_ast: vy_ast.Module):
+        acc = [sha256sum(module_ast.full_source_code)]
+        for s in module_ast.get_children((vy_ast.Import, vy_ast.ImportFrom)):
+            info = s._metadata["import_info"]
+
+            if info.compiler_input.path.suffix in (".vyi", ".json"):
+                # NOTE: this needs to be redone if interfaces can import other interfaces
+                acc.append(info.compiler_input.sha256sum)
+            else:
+                acc.append(self._calculate_integrity_sum_r(info.parsed))
+
+        return sha256sum("".join(acc))
+
+    def _resolve_imports_r(self, module_ast: vy_ast.Module):
+        if id(module_ast) in self.seen:
+            return
+        with self.graph.enter_path(module_ast):
+            for node in module_ast.body:
+                if isinstance(node, vy_ast.Import):
+                    self._handle_Import(node)
+                elif isinstance(node, vy_ast.ImportFrom):
+                    self._handle_ImportFrom(node)
+        self.seen.add(id(module_ast))
+
+    def _handle_Import(self, node: vy_ast.Import):
+        # import x.y[name] as y[alias]
+
+        alias = node.alias
+
+        if alias is None:
+            alias = node.name
+
+        # don't handle things like `import x.y`
+        if "." in alias:
+            msg = "import requires an accompanying `as` statement"
+            suggested_alias = node.name[node.name.rfind(".") :]
+            hint = f"try `import {node.name} as {suggested_alias}`"
+            raise StructureException(msg, node, hint=hint)
+
+        self._add_import(node, 0, node.name, alias)
+
+    def _handle_ImportFrom(self, node: vy_ast.ImportFrom):
+        # from m.n[module] import x[name] as y[alias]
+        alias = node.alias or node.name
```

Comment:
inconsistent with `_handle_Import` 

```python
alias = node.alias

if alias is None:
    alias = node.name
```

Reply:
i can change it, but that's how it was before. `node.alias == ""` is not an allowable state by the parser.

Reply:
let's update please, this is minor and safe

Reply:
addressed in: https://github.com/vyperlang/vyper/pull/4229/commits/1206a8defe992df4484c27802e201f264dc182c0